Add combined function bend element.

cemitch99 opened the pull request:
Includes the linear map for a combined-function bend, consisting of a sector bend with a nonzero quadrupole field component.

Related to #374

- [x] Add element particle push.
- [x] Add benchmark example.
- [x] Add Python equivalents.
- [x] Update documentation.
  - [x] link in TOC tree of `examples.rst`

cemitch99 committed:
Examples for 3D space charge benchmarking

- Modified the initial beam size in the IOTA lens benchmark example.
- Added 2 benchmarks of 3D space charge for initial testing.
- Add documentation for 2 benchmarks with space charge.
- Add a benchmark example with space charge and periodic s-dependent focusing.
- Added an s-dependent example using a Kurth beam without space charge.
- Modified tolerance for IOTA lens benchmark example.
  Reduced tolerance to account for smaller initial beam size and
  improved preservation of invariants of motion.
- Modified tolerances of space charge examples to allow CI tests to
  pass when space charge is not active.

- Modified tolerance for space charge examples.
  These should fail unless space charge is turned on.

cemitch99 committed:
Update input_kurth_10nC.in

Selected numerical values for amr.n_cell, lattice.nslice, and geometry.prob_relative.

cemitch99 committed:
Add combined function bend element.

cemitch99 committed:
Delete input_kurth_10nC.in

This is not part of this PR.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update documentation.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 committed:
Update All.H

Adjust line format

ax3l committed:
CFbend: Formatting

ax3l commented:
This looks great, thank you!

I fixed the formatting a bit and pushed the changes. Do you like to add a benchmark in this PR or shall we do a follow-up? :)

cemitch99 committed:
Add benchmark example.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

cemitch99 commented:
> This looks great, thank you!
> 
> I fixed the formatting a bit and pushed the changes. Do you like to add a benchmark in this PR or shall we do a follow-up? :)

I added a benchmark.

cemitch99 committed:
Add link in TOC of examples.rst.

ax3l commented on examples/cfbend/README.rst:
```diff
@@ -0,0 +1,47 @@
+.. _examples-cfbend:
+
+Combined Function Bend
+=======================
```
```suggestion
Combined Function Bend
======================
```

ax3l commented on examples/cfbend/README.rst:
```diff
@@ -0,0 +1,47 @@
+parameters are based a single CSBEND element appearing in the ELEGANT input file for the ALS-U lattice.
+
+The beam parameters are based on:
+C. Steier et al, "Status of the Conceptual Design of ALS-U", IPAC2017, WEPAB104 (2017).
```
I only found this in google scholar:
```suggestion
C. Steier et al, "Status of the Conceptual Design of ALS-U", MEDSI2018, TUPH14, `DOI:10.18429/JACoW-MEDSI2018-TUPH14 <https://doi.org/10.18429/JACoW-MEDSI2018-TUPH14>`__ (2018).
```

ax3l replied:
Oh, this one? :)
```suggestion
C. Steier et al, "Status of the Conceptual Design of ALS-U", IPAC2017, WEPAB104, `DOI:10.18429/JACoW-IPAC2017-WEPAB104 <https://doi.org/10.18429/JACoW-IPAC2017-WEPAB104>`__ (2017).
```

ax3l commented on examples/cfbend/input_cfbend.in:
```diff
@@ -0,0 +1,46 @@
+# Particle Beam(s)
+###############################################################################
```
```suggestion
###############################################################################
# Particle Beam(s)
###############################################################################
```

ax3l commented on examples/cfbend/input_cfbend.in:
```diff
@@ -0,0 +1,46 @@
+cfbend1.type = cfbend
+cfbend1.ds = 0.5       # projected length 0.5 m, angle 3.76 deg
+cfbend1.rc = 7.613657587094493   # bending radius [m]
+cfbend1.k = -7.057403   #(upright) quadrupole component [m^(-2)]
```
```suggestion
cfbend1.k = -7.057403   # (upright) quadrupole component [m^(-2)]
```

ax3l commented on examples/cfbend/run_cfbend.py:
```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022-2023 ImpactX contributors
+# Authors: Marco Garten, Axel Huebl, Chad Mitchell
```
Probably copied?
```suggestion
# Authors: Chad Mitchell, Axel Huebl
```
Feel free to add Marco if he contributed to this example.

ax3l commented on examples/cfbend/run_cfbend_madx.py:
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022-2023 ImpactX contributors
+# Authors: Marco Garten, Axel Huebl, Chad Mitchell
```
As above
```suggestion
# Authors: Chad Mitchell, Axel Huebl
```

ax3l committed:
Cleaning: Format, Comments

ax3l committed:
Add DOI

ax3l reviewed (approved):
Looks great, thank you!
